release/microbenchmark/experimental/accelerated_dag_gpu_microbenchmark.py

```diff
@@ -0,0 +1,353 @@
+
+
+@ray.remote(num_gpus=1)
+class NcclWorker:
+    def __init__(self, world_size, rank, comm_id):
+        from ray.air._internal import torch_utils
+
+        self.device = torch_utils.get_devices()[0]
+        self.world_size = world_size
+        self.rank = rank
+
+        torch.distributed.init_process_group(
+            backend="nccl", world_size=world_size, rank=rank
+        )
+
+    def _send(self, buf, num_el, rank):
+        torch.distributed.send(buf, rank)
+
+    def _recv(self, buf, num_el, rank):
+        torch.distributed.recv(buf, rank)
+
+    def do_send_recv(self, shape, dtype):
+        other_rank = (self.rank + 1) % self.world_size
+
+        def _run():
+
+            if self.rank == 0:
+                i = np.random.randint(100)
+                input_buffer = torch.ones(shape, dtype=dtype, device=self.device) * i
+                self._send(input_buffer, input_buffer.numel(), other_rank)
+            else:
+                input_buffer = torch.zeros(shape, dtype=dtype, device=self.device)
+                self._recv(input_buffer, input_buffer.numel(), other_rank)
+
+            torch.cuda.synchronize()
+
+        timeit("exec_nccl_gpu", _run)
+
+
+def exec_ray_dag(
+    label, sender, receiver, use_nccl=False, use_adag=True
+):
+    # Test torch.Tensor sent between actors.
+    with InputNode() as inp:
+        dag = sender.send.bind(SHAPE, DTYPE, inp)
+
+        if use_adag:
+            dag = dag.with_type_hint(
+                TorchTensorType(SHAPE, DTYPE, transport="nccl" if use_nccl else None)
+            )
+
+        dag = receiver.recv.bind(dag)
+
+    if use_adag:
+        dag = dag.experimental_compile(buffer_size_bytes=int(SHAPE[0] * 3))
+
+        def _run():
+            i = np.random.randint(100)
+            output_channel = dag.execute(i)
+            # TODO(swang): Replace with fake ObjectRef.
+            result = output_channel.begin_read()
+            assert result == (i, SHAPE, DTYPE)
+            output_channel.end_read()
+
+    else:
+
+        def _run():
+            i = np.random.randint(100)
+            result = ray.get(dag.execute(i))
+            assert result == (i, SHAPE, DTYPE)
+
+    results = timeit(label, _run)
+
+    if use_adag:
+        dag.teardown()
+
+    # Workaround for Ray bug in reusing GPUs too quickly.
+    # See https://github.com/ray-project/ray/issues/44821.
+    ray.kill(sender)
+    ray.kill(receiver)
+    time.sleep(1)
+
+    return results
+
+
+def exec_ray_dag_ipc(label, sender, receiver, use_nccl=False):
+    # Test torch.Tensor sent between actors.
+    with InputNode() as inp:
+        dag = sender.send.bind(SHAPE, DTYPE, inp)
+        dag = receiver.recv.bind(
+            dag,
+            SHAPE[0] * DTYPE.itemsize,
+            SHAPE,
+            nccl_util.TORCH_NUMPY_DTYPE_MAP[DTYPE],
+        )
+
+    compiled_dag = dag.experimental_compile(buffer_size_bytes=int(SHAPE[0] * 3))
+    ok = [True]
```

Comment:
Why a list?

Reply:
Ah it's so that the lambda can close over the list and update its value.